transcom/mymove
# B-17616-TOO-Can-Edit-SIT-Entry-Date

## [API & UI - TOO can Edit First Day of SIT Date](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A844136&RoomContext=TeamRoom%3A808731&concept=TeamRoom)
## [MB-16666: TOO needs to be able to edit first day of SIT](https://www13.v1host.com/USTRANSCOM38/Epic.mvc/Summary?oidToken=Epic%3A834520)

## Summary

This is the only story for the feature where a TOO can edit the first day of SIT. 
This story involved:
- adding a PATCH endpoint to the GHC API to update a SIT entry date for a single service item (`/service-item/{serviceItemID}/entry-date-update`)
- activating a non-implemented GET endpoint in the GHC API to retrieve a service item from the database (`/move-task-orders/{moveTaskOrderID}/service-items/{mtoServiceItemID}`)
- adding a custom UI component in the frontend for displaying a modal and submitting the new SIT entry date
- updating the database with new value and then updating Redux state for the selected service item
- tests for both frontend and backend covering added code

Some things to note:
The displayed "SIT (STORAGE IN TRANSIT)" box at the top of the Move Task Order component does contain some SIT data that is displaying values from another area of state called `currentSit` and has not been updated due to it being out of scope of this story. This could be an enhancement for the future, but when the service item is updated successfully, the state does change for that specific service item in Redux.

ALSO, I had to implement the comment `/* istanbul ignore next */` in the MoveTaskOrder component. It was found that the functions that were being considered for test coverage could not be exported and thus could not be tested. The only way around that would be to move those functions into their own file and import that into the MoveTaskOrder component and the MoveTaskOrder.test file. I only added the comment for functions that could not be exported to satisfy the `client_test_coverage` requirements. I understand this should be used sparingly, but we should consider potentially moving large blocks of functions like that that can't be exported to their own file for future implementations or enhancements.

### Setup to Run the Code

- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

1. Log into MilMove Office as a TOO
2. Find a Move that contains SIT service items
3. In the Move Task Order page, scroll down to see the list of service items
4. You should see an "Edit" button next to origin/destination 1st Day SIT service items
5. Click on the edit button and change the SIT entry date, enter office remarks, and click Save button
6. You should see a success banner at the top of the page and scroll down to see the SIT entry date has changed for the respective service item

### Frontend

There is an existing console error that was present prior to developing this story when updating any service item (like when updating the address) that is due to some undefined values in a shipment object somewhere. This was not addressed in this story due to it being out of scope.
- [x] All tests created for new implementation are passing

### Backend

- [x] All tests created for new implementation are passing

## Screenshots
